### PR TITLE
FEATURE: Introduce personal_message_enabled_groups setting

### PR DESCRIPTION
**Pull request by martin-brennan**
This will replace `enable_personal_messages` and
`min_trust_to_send_messages`, this commit introduces
the setting `personal_message_enabled_groups`
and uses it in all places that `enable_personal_messages`
and `min_trust_to_send_messages` currently apply.

A migration is included to set `personal_message_enabled_groups`
based on the following rules:

* If `enable_personal_messages` was false, then set
  `personal_message_enabled_groups` to `3`, which is
  the staff auto group
* If `min_trust_to_send_messages` is not default (1)
  and the above condition is false, then set the
  `personal_message_enabled_groups` setting to
  the appropriate auto group based on the trust level
* Otherwise just set `personal_message_enabled_groups` to
  11 which is the TL1 auto group

After follow-up PRs to plugins using these old settings, we will be
able to drop the old settings from core, in the meantime I've added
 DEPRECATED notices to their descriptions and added them
to the deprecated site settings list.

This commit also introduces a `_map` shortcut method definition
for all `group_list` site settings, e.g. `SiteSetting.personal_message_enabled_groups`
also has `SiteSetting.personal_message_enabled_groups_map` available,
which automatically splits the setting by `|` and converts it into
an array of integers.

### CODE REVIEW

File: app/assets/javascripts/discourse/app/models/site.js
Comment (martin-brennan) [inline app/assets/javascripts/discourse/app/models/site.js:143]: I am not sure if site is the best place for this, feel free to suggest somewhere else.

File: app/models/site_setting.rb
Comment (martin-brennan) [inline app/models/site_setting.rb:146]: A follow-up PR can be done to replace other places in the codebase that do this splitting pattern. Alternatively, I did think of getting all the `group_list` settings and dynamically creating methods for them to do this split/map e.g. `personal_message_enabled_groups_map`, so we can do `SiteSetting.personal_message_enabled_groups_map`

File: lib/guardian.rb
Comment (martin-brennan) [inline lib/guardian.rb:406]: There are many such places in the backend where I need to leave `enable_personal_messages` in place, if the condition was removed then plugin tests would fail since they are not setting up user auto groups based on trust level correctly. Once I fix the plugins and update them to remove usages of `enable_personal_messages` and `min_trust_to_send_messages` then we will be good to remove this from core.

File: spec/lib/post_action_creator_spec.rb
Comment (martin-brennan) [inline spec/lib/post_action_creator_spec.rb:10]: I chose to selectively apply this to the test that need it, I found automatically doing it for all tests e.g. in the user fabricator was a disaster because there are many tests relying on specific group counts, and we also don't want this huge overhead for tests that don't need it.

File: app/assets/javascripts/discourse/app/models/site.js
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/models/site.js:151]: can this be empty? If yes we should `filter(Boolean)`

```javascript
"".split("|") // [""]
````

File: app/assets/javascripts/discourse/tests/fixtures/session-fixtures.js
Comment (jjaffeux) [inline app/assets/javascripts/discourse/tests/fixtures/session-fixtures.js:37]: 😿

Comment (martin-brennan) [reply app/assets/javascripts/discourse/tests/fixtures/session-fixtures.js:37]: I realised I can probably at least cut this down to the essentials. It more reflects reality -- in most cases every user will have _at least_ TL0 and TL1 in this array, so it makes sense to include it in the fixture.

File: app/assets/javascripts/discourse/app/models/site.js
Comment (martin-brennan) [reply app/assets/javascripts/discourse/app/models/site.js:151]: No it should never be empty, I will need to make sure the site setting has validation for that

File: app/assets/javascripts/discourse/app/components/topic-footer-buttons.js
Comment (ZogStriP) [inline app/assets/javascripts/discourse/app/components/topic-footer-buttons.js:44]: Shoudn't this also depend on `canSendPms` like you did for `canArchive`?

File: app/assets/javascripts/discourse/app/components/topic-footer-buttons.js
Comment (ZogStriP) [inline app/assets/javascripts/discourse/app/components/topic-footer-buttons.js:49]: It's not obvious at all that we need to listen to the current user's groups to have an accurate result for `allowPersonalMessages` 🤔 

How about we make it more explicit and change the signature of `allowPersonalMessages` to take these parameters `(isStaff, groups)`?

Comment (ZogStriP) [reply app/assets/javascripts/discourse/app/components/topic-footer-buttons.js:49]: Sometimes you're listening for both `currentUser.staff` and `currentUser.groups.[]` and other times, only for `currentUser.groups.[]`.

Comment (martin-brennan) [reply app/assets/javascripts/discourse/app/components/topic-footer-buttons.js:49]: I will listen to both instead :+1: And yeah good idea to add the additional parameter to make it more explicit.

Comment (martin-brennan) [reply app/assets/javascripts/discourse/app/components/topic-footer-buttons.js:49]: Oh I just reread:

> How about we make it more explicit and change the signature of allowPersonalMessages to take these parameters (isStaff, groups)?

This is more confusing to me...the function is checking whether the current user (if there is one) can send PMs based on the groups they are a member of, using `personal_message_enabled_groups` as a reference. I think the params as is are fine, but I will fix the computeds to check for `currentUser.staff` too

File: app/assets/javascripts/discourse/app/models/site.js
Comment (martin-brennan) [reply app/assets/javascripts/discourse/app/models/site.js:151]: Ok this is done

File: app/assets/javascripts/discourse/app/components/topic-footer-buttons.js
Comment (ZogStriP) [reply app/assets/javascripts/discourse/app/components/topic-footer-buttons.js]: I guess my point was that I find it hard to know that whenever you call `allowPersonalMessages` you have to listen to both `currentUser.staff` and `currentUser.groups.[]`.

**Comment by jjaffeux** [comment]
Maybe we move allowPersonalMessages to current user service, in your code it's only code with current user as a param afaict. This way we can compute on current user groups and other things and don't have to worry about this.

If the need for a different user than current would show up, we could think of other solutions.

**Comment by martin-brennan** [comment]
> Maybe we move allowPersonalMessages to current user service, in your code it's only code using current user as a param afaict. This way we can compute on current user groups and other things and don't have to worry about this.

Oh of course that will be better, its a fine tradeoff for now.

File: app/assets/javascripts/discourse/app/components/topic-footer-buttons.js
Comment (martin-brennan) [reply app/assets/javascripts/discourse/app/components/topic-footer-buttons.js]: Yes I see now @ZogStriP . I am going to move this into current user service anyway at Joffrey's suggestion :+1:

File: spec/lib/admin_user_index_query_spec.rb
Comment (ZogStriP) [inline spec/lib/admin_user_index_query_spec.rb:84]: Might be clearer if we used the enum rather than 5?

**Review by ZogStriP — APPROVED** [review]

File: app/assets/javascripts/discourse/app/components/topic-footer-buttons.js
Comment (ZogStriP) [reply app/assets/javascripts/discourse/app/components/topic-footer-buttons.js]: My bad, I wasn't clear enough. Thankfully @jjaffeux explained my idea better 👍

File: spec/lib/admin_user_index_query_spec.rb
Comment (martin-brennan) [reply spec/lib/admin_user_index_query_spec.rb:84]: Nice didn't know that was a thing, will fix.

**Comment by martin-brennan** [comment]
Spec failures unrelated

File: app/models/user.rb
Comment (martin-brennan) [inline app/models/user.rb:415]: @tgxworld the second change, checks if the user is in any of the passed in group IDs based on their `GroupUser` relation via:

https://github.com/discourse/discourse/blob/15119b524d13f9842836aa4dcc53672569b3283c/app/models/user.rb#L87-L87

Usage with site setting change:

```ruby
@user.in_any_groups?(SiteSetting.group_setting_map(:personal_message_enabled_groups))
```

I guess the main improvement we could do here is a stripped down scope that just gets the group IDs and names?

File: db/migrate/20220825054405_fill_personal_message_enabled_groups_based_on_deprecated_settings.rb
Comment (martin-brennan) [inline db/migrate/20220825054405_fill_personal_message_enabled_groups_based_on_deprecated_settings.rb:4]: @tgxworld would you mind looking over the migration too, another set of eyes is always helpful on ones like these :eyes:

File: app/models/site_setting.rb
Comment (tgxworld) [reply app/models/site_setting.rb:146]: > Maybe we can be cleverer and for all group_list settings have a version of it with _map on the end, e.g. SiteSetting.personal_message_enabled_groups_map?

Yea I like this idea :+1:

File: app/models/user.rb
Comment (tgxworld) [inline app/models/user.rb:416]:
```suggestion
    group_ids.include?(Group::AUTO_GROUPS[:everyone]) || (group_ids & group_users.pluck(&:group_id)).any?
```

Calling `#groups` will load the AR records which is not really necessary.

File: app/models/site_setting.rb
Comment (martin-brennan) [reply app/models/site_setting.rb:146]: Okay cool I will change it over, should be easy @tgxworld and saves verbosity

File: app/models/user.rb
Comment (martin-brennan) [reply app/models/user.rb:416]: Sure will strip that back, I will double check that this doesn't cause multiple queries, and if so I will memoize it for the user in a different method e.g. `belonging_to_group_ids` or something

File: db/migrate/20220825054405_fill_personal_message_enabled_groups_based_on_deprecated_settings.rb
Comment (tgxworld) [reply db/migrate/20220825054405_fill_personal_message_enabled_groups_based_on_deprecated_settings.rb:4]: Something I noticed is that we don't seem to be disallowing write access to the old site setting in this PR. I think we need to deprecate `SiteSetting.enable_personal_messages` and `SiteSetting.min_trust_to_send_messages_raw` and create some sort of map from `SiteSetting.personal_messages_enabled_groups` to the old site settings?

Comment (martin-brennan) [reply db/migrate/20220825054405_fill_personal_message_enabled_groups_based_on_deprecated_settings.rb:4]: Ah very good point, that's what we did with the whitelist/allowlist stuff here, will have to do that for these settings too:

https://github.com/discourse/discourse/blob/15119b524d13f9842836aa4dcc53672569b3283c/app/models/site_setting.rb#L271-L280

Comment (martin-brennan) [reply db/migrate/20220825054405_fill_personal_message_enabled_groups_based_on_deprecated_settings.rb:4]: Hmm actually it doesn't really map very well, and deprecating + throwing an error when changing the setting breaks tons of specs, and I don't really want to do a fragile mapping between these two settings and the single new one. There aren't many plugins relying on this, and none of them are ours :shrug: Don't want to hold this up any longer so just adding the deprecations and being done with it.

Comment (martin-brennan) [reply db/migrate/20220825054405_fill_personal_message_enabled_groups_based_on_deprecated_settings.rb:4]: Never mind, found the deprecate site settings class :)

**Pull request merged**
